### PR TITLE
JP-3768: Move outlier detection median computers to stcal

### LINKED ISSUE #8838: Outlier detection median computers to stcal
_Issue [JP-3768](https://jira.stsci.edu/browse/JP-3768) was created on JIRA by [Ned Molter](https://jira.stsci.edu/secure/ViewProfile.jspa?name=emolter):_

New memory- and runtime-optimized infrastructure for computing the median during outlier has recently been added to the jwst repository.  This should be moved to stcal for use in Roman and perhaps elsewhere.

### PR DESCRIPTION
**Pull request by emolter**
Resolves [JP-3768](https://jira.stsci.edu/browse/JP-3768)

Closes #8838

This PR ports the median calculation machinery into stcal for use by other missions.  See https://github.com/spacetelescope/stcal/pull/292 for the related changes in stcal.

## Tasks
- [x] **request a review from someone specific**, to avoid making the maintainers review every PR
- [x] add a build milestone, i.e. `Build 11.3` (use the [latest build](https://github.com/spacetelescope/jwst/milestones) if not sure)
- [ ] Does this PR change user-facing code / API? (if not, label with `no-changelog-entry-needed`)
  - [x] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types) 
  - [x] update or add relevant tests
  - [ ] update relevant docstrings and / or `docs/` page
  - [x] [start a regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) and include a link to the running job ([click here for instructions](https://github.com/spacetelescope/RegressionTests/blob/main/docs/running_regression_tests.md))
    - [ ] Do truth files need to be updated ("okified")?
      - [ ] **after the reviewer has approved these changes**, run `okify_regtests` to update the truth files
- [ ] if a JIRA ticket exists, [make sure it is resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
- ``changes/<PR#>.docs.rst``
- ``changes/<PR#>.stpipe.rst``
- ``changes/<PR#>.datamodels.rst``
- ``changes/<PR#>.scripts.rst``
- ``changes/<PR#>.fits_generator.rst``
- ``changes/<PR#>.set_telescope_pointing.rst``
- ``changes/<PR#>.pipeline.rst``

## stage 1
- ``changes/<PR#>.group_scale.rst``
- ``changes/<PR#>.dq_init.rst``
- ``changes/<PR#>.emicorr.rst``
- ``changes/<PR#>.saturation.rst``
- ``changes/<PR#>.ipc.rst``
- ``changes/<PR#>.firstframe.rst``
- ``changes/<PR#>.lastframe.rst``
- ``changes/<PR#>.reset.rst``
- ``changes/<PR#>.superbias.rst``
- ``changes/<PR#>.refpix.rst``
- ``changes/<PR#>.linearity.rst``
- ``changes/<PR#>.rscd.rst``
- ``changes/<PR#>.persistence.rst``
- ``changes/<PR#>.dark_current.rst``
- ``changes/<PR#>.charge_migration.rst``
- ``changes/<PR#>.jump.rst``
- ``changes/<PR#>.clean_flicker_noise.rst``
- ``changes/<PR#>.ramp_fitting.rst``
- ``changes/<PR#>.gain_scale.rst``

## stage 2
- ``changes/<PR#>.assign_wcs.rst``
- ``changes/<PR#>.badpix_selfcal.rst``
- ``changes/<PR#>.msaflagopen.rst``
- ``changes/<PR#>.nsclean.rst``
- ``changes/<PR#>.imprint.rst``
- ``changes/<PR#>.background.rst``
- ``changes/<PR#>.extract_2d.rst``
- ``changes/<PR#>.master_background.rst``
- ``changes/<PR#>.wavecorr.rst``
- ``changes/<PR#>.srctype.rst``
- ``changes/<PR#>.straylight.rst``
- ``changes/<PR#>.wfss_contam.rst``
- ``changes/<PR#>.flatfield.rst``
- ``changes/<PR#>.fringe.rst``
- ``changes/<PR#>.pathloss.rst``
- ``changes/<PR#>.barshadow.rst``
- ``changes/<PR#>.photom.rst``
- ``changes/<PR#>.pixel_replace.rst``
- ``changes/<PR#>.resample_spec.rst``
- ``changes/<PR#>.residual_fringe.rst``
- ``changes/<PR#>.cube_build.rst``
- ``changes/<PR#>.extract_1d.rst``
- ``changes/<PR#>.resample.rst``

## stage 3
- ``changes/<PR#>.assign_mtwcs.rst``
- ``changes/<PR#>.mrs_imatch.rst``
- ``changes/<PR#>.tweakreg.rst``
- ``changes/<PR#>.skymatch.rst``
- ``changes/<PR#>.exp_to_source.rst``
- ``changes/<PR#>.outlier_detection.rst``
- ``changes/<PR#>.tso_photometry.rst``
- ``changes/<PR#>.stack_refs.rst``
- ``changes/<PR#>.align_refs.rst``
- ``changes/<PR#>.klip.rst``
- ``changes/<PR#>.spectral_leak.rst``
- ``changes/<PR#>.source_catalog.rst``
- ``changes/<PR#>.combine_1d.rst``
- ``changes/<PR#>.ami.rst``

## other
- ``changes/<PR#>.wfs_combine.rst``
- ``changes/<PR#>.white_light.rst``
- ``changes/<PR#>.cube_skymatch.rst``
- ``changes/<PR#>.engdb_tools.rst``
- ``changes/<PR#>.guider_cds.rst``
</details>

### CODE REVIEW

**Comment by emolter** [comment]
starting [regression tests here](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1758/)

**Review by melanieclarke — APPROVED** [review]
LGTM, pending merge of the stcal PR.  I like the new append and evaluate methods much better than the old helper functions!

File: jwst/outlier_detection/utils.py
Comment (braingram) [inline jwst/outlier_detection/utils.py:13]: Is nanmedian3D used?

File: jwst/outlier_detection/utils.py
Comment (braingram) [inline jwst/outlier_detection/utils.py:99]: Somewhat unrelated to this PR (so feel free to say "make a ticket") but doesn't the above line modify the input model data?

File: jwst/outlier_detection/utils.py
Comment (braingram) [reply jwst/outlier_detection/utils.py:13]: Nevermind. It's in `create_cube_median`.

File: jwst/outlier_detection/utils.py
Comment (emolter) [reply jwst/outlier_detection/utils.py:99]: I think you mean a different line.  How does appending `drizzled_model.data` to the computer modify it?

I'm a little confused about the "never modify the input data" rule.  Does that mean that for a step like outlier detection where you have a ModelLibrary input and output, you should never call `modify=True` when shelving?

I'd say let's put any changes into a different refactor or PR, unless you have concerns w.r.t. the functions that are going into stcal.

Comment (braingram) [reply jwst/outlier_detection/utils.py:99]: Line 98 is that one that can modify the input data. I opened an issue: https://github.com/spacetelescope/jwst/issues/8845

The "never modify the input data" rule is certainly beyond the scope of this PR (and doesn't really apply for most if not all of image 3).

No concerns about this PR. I just brought it up since I noticed it while working on the corresponding romancal PR and wanted to check if I was mistaken.

**Comment by emolter** [comment]
https://github.com/spacetelescope/stcal/pull/292 is approved and merged.  What changes if any do you recommend to `pyproject.toml` before merging this one?  Or do we need to wait for an stcal release? @braingram @melanieclarke

**Comment by braingram** [comment]
> [spacetelescope/stcal#292](https://github.com/spacetelescope/stcal/pull/292) is approved and merged. What changes if any do you recommend to `pyproject.toml` before merging this one? Or do we need to wait for an stcal release? @braingram @melanieclarke

Roman folks typically do something like:
```
#    "stcal>=1.8.0,<1.9.0",
    "stcal @ git+https://github.com/spacetelescope/stcal.git@main",
```
I think that should work here. We'll of course need a release of stcal before any release of jwst.

**Comment by emolter** [comment]
started a (hopefully final) round of [regression tests](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1761/) to ensure the correct stcal version gets picked up and works properly

**Comment by braingram** [comment]
> started a (hopefully final) round of [regression tests](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1761/) to ensure the correct stcal version gets picked up and works properly

The jenkins runs are a bit of a lie for changes like this. They always use dev requirements for some packages:
https://github.com/spacetelescope/jwst/blob/72483584eebe7408bdf9f91f11a3690af136db6d/JenkinsfileRT#L76
one of which is stcal:
https://github.com/spacetelescope/jwst/blob/72483584eebe7408bdf9f91f11a3690af136db6d/requirements-dev-st.txt#L5
So anything listed in pyproject.toml for stcal is ignored for jenkins.

**Comment by emolter** [comment]
> The jenkins runs are a bit of a lie for changes like this. They always use dev requirements for some packages:

Oh, right, I think I already knew that.  Nevertheless, the tests should pass now that the merge into stcal happened, right?

**Comment by braingram** [comment]
> Oh, right, I think I already knew that. Nevertheless, the tests should pass now that the merge into stcal happened, right?

They should. Also, I don't believe the github actions regtests by default install `requirements-dev-st.txt` (which I think makes sense but I'm happy to discuss this with anyone interested) so the update to pyproject.toml should also:
- fix the github CI for this PR (and subsequent PRs once this is merged)
- fix the github regression tests
- fix installs for users/developers who `pip install` from the dev branch
- provide a more standard description of the actual requirements

**Pull request merged**
